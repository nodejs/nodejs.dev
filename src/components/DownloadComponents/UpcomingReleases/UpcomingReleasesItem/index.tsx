import React from 'react';
import { FormattedMessage } from 'react-intl';
import classnames from 'classnames';
import { ReactComponent as HexagonFilled } from '../../../../images/icons/hexagon-filled.svg';
import { ReactComponent as HexagonOutline } from '../../../../images/icons/hexagon-outline.svg';
import { UpcomingReleaseData } from '../../../../types';
import styles from './index.module.scss';

type Props = UpcomingReleaseData;

const UpcomingReleasesItem = ({
  releaseType,
  releaseDate,
  alreadyReleased,
}: Props): JSX.Element => {
  const Image = alreadyReleased ? HexagonFilled : HexagonOutline;
  const Released = alreadyReleased ? <FormattedMessage id="components.UpcomingReleasesItem" /> : '';

  const className = classnames({
    [styles.itemCurrent]: releaseType === 'Current',
    [styles.itemLts]: releaseType === 'LTS',
    [styles.itemMaintenance]: releaseType === 'Maintenance',
    [styles.itemEndoflife]: releaseType === 'End-of-life',
    [styles.itemToBeReleased]: !alreadyReleased,
  });

  return (
    <div className={className}>
      <Image />
      <p className={styles.releaseTitle}>{releaseType}</p>
      <p className={styles.releaseDate}>
        {Released} {releaseDate}
      </p>
    </div>
  );
};

export default UpcomingReleasesItem;
