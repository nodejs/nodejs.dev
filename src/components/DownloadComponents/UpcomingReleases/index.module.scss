.upcoming-releases {
  width: 100%;
}

.react-tabs {
  -webkit-tap-highlight-color: transparent;

  &__tab-list {
    border-bottom: 1px solid #d7d8df;
    display: flex;
    margin: 0;
    overflow-x: auto;
    padding: 0;
    text-align: center;
  }

  &__tab {
    cursor: pointer;
    display: inline-block;
    list-style: none;
    padding: 7px 20px;

    &--selected {
      border-bottom: 4px solid var(--brand5);

      &:focus {
        outline: none;
      }
    }
  }

  &__tab-panel {
    display: none;
    padding: 70px 0;

    &--selected {
      display: block;
    }
  }
}
