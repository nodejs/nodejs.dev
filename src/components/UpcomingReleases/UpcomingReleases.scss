.upcoming-releases {
  width: 100%;
  .react-tabs {
    -webkit-tap-highlight-color: transparent;

    &__tab-list {
      margin: 0;
      padding: 0;
      text-align: center;
      display: flex;
      border-bottom: 1px solid #d7d8df;
    }

    &__tab {
      display: inline-block;
      list-style: none;
      padding: 7px 20px;
      cursor: pointer;

      &--selected {
        border-bottom: 4px solid var(--brand5);
      }
    }

    &__tab-panel {
      display: none;
      padding: 70px 0;

      &--selected {
        display: block;
      }
    }
  }

  &__panel {
    display: grid;
    grid-template-columns: repeat(4, 1fr);
    grid-template-rows: 1fr;
    grid-column-gap: 0px;
    grid-row-gap: 0px;
    text-align: center;
  }
}
