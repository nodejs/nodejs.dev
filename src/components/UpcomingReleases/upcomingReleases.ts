export type ReleaseTypes = 'current' | 'lts' | 'maintenance' | 'endoflife';

export type UpcomingRelease = {
  releaseDate: string;
  releaseType: ReleaseTypes;
  alreadyReleased: boolean;
};

export const RELEASE_TYPES: { [key in ReleaseTypes]: string } = {
  current: 'Current',
  lts: 'LTS',
  maintenance: 'Maintenance Release',
  endoflife: 'End-of-life Release',
};

export const NODE_12_RELEASES: UpcomingRelease[] = [
  {
    alreadyReleased: true,
    releaseDate: '2019-04-23',
    releaseType: 'current',
  },
  {
    alreadyReleased: true,
    releaseDate: '2019-10-21',
    releaseType: 'lts',
  },
  {
    alreadyReleased: true,
    releaseDate: '2020-10-20',
    releaseType: 'maintenance',
  },
  {
    alreadyReleased: false,
    releaseDate: '2022-04-30',
    releaseType: 'endoflife',
  },
];

export const NODE_13_RELEASES: UpcomingRelease[] = [
  {
    alreadyReleased: true,
    releaseDate: '2019-10-22',
    releaseType: 'current',
  },
  {
    alreadyReleased: true,
    releaseDate: '2020-04-01',
    releaseType: 'maintenance',
  },
  {
    alreadyReleased: true,
    releaseDate: '2020-06-01',
    releaseType: 'endoflife',
  },
];

export const NODE_14_RELEASES: UpcomingRelease[] = [
  {
    alreadyReleased: true,
    releaseDate: '2020-04-21',
    releaseType: 'current',
  },
  {
    alreadyReleased: true,
    releaseDate: '2020-10-20',
    releaseType: 'lts',
  },
  {
    alreadyReleased: false,
    releaseDate: '2021-10-19',
    releaseType: 'maintenance',
  },
  {
    alreadyReleased: false,
    releaseDate: '2023-04-30',
    releaseType: 'endoflife',
  },
];

export const NODE_15_RELEASES: UpcomingRelease[] = [
  {
    alreadyReleased: true,
    releaseDate: '2020-10-21',
    releaseType: 'current',
  },
  {
    alreadyReleased: false,
    releaseDate: '2021-04-01',
    releaseType: 'maintenance',
  },
  {
    alreadyReleased: false,
    releaseDate: '2021-06-01',
    releaseType: 'endoflife',
  },
];
