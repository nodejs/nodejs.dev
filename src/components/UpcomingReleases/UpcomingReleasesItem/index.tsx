import React from 'react';
import classnames from 'classnames';
import hexagonFilled from '../../../images/icons/hexagon-filled.svg';
import hexagonOutline from '../../../images/icons/hexagon-outline.svg';
import { UpcomingRelease, RELEASE_TYPES } from '../upcomingReleases';
import './UpcomingReleasesItem.scss';

type Props = UpcomingRelease;

export default function UpcomingReleasesItem({
  releaseType,
  releaseDate,
  alreadyReleased,
}: Props): JSX.Element {
  const image = alreadyReleased ? hexagonFilled : hexagonOutline;
  const className = classnames(`upcoming-releases__item--${releaseType}`, {
    'upcoming-releases__item--to-be-released': !alreadyReleased,
  });

  return (
    <div className={className}>
      <img src={image} alt="hexagon icon" />
      <p className="release-title">{RELEASE_TYPES[releaseType]}</p>
      <p className="release-date">
        {alreadyReleased ? 'Released' : 'To be released'} {releaseDate}
      </p>
    </div>
  );
}
