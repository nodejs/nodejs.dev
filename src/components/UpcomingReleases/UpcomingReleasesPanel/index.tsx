import React from 'react';
import UpcomingReleasesItem from '../UpcomingReleasesItem';
import { UpcomingRelease } from '../upcomingReleases';

type Props = {
  releases: UpcomingRelease[];
};

export default function UpcomingReleasesPanel({
  releases,
}: Props): JSX.Element {
  return (
    <div className="upcoming-releases__panel">
      {releases.map(release => (
        <UpcomingReleasesItem
          key={release.releaseType}
          releaseDate={release.releaseDate}
          releaseType={release.releaseType}
          alreadyReleased={release.alreadyReleased}
        />
      ))}
    </div>
  );
}
