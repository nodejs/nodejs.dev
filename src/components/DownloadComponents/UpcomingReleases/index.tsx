import React from 'react';
import { Tab, TabList, TabPanel, Tabs } from 'react-tabs';
import UpcomingReleasesPanel from './UpcomingReleasesPanel';
import { UpcomingRelease } from '../../../types';
import styles from './index.module.scss';

interface Props {
  upcomingReleases: UpcomingRelease[];
}

const UpcomingReleases = ({ upcomingReleases }: Props): JSX.Element => (
  <div className={styles.upcomingReleases}>
    <Tabs className={styles.reactTabs}>
      <TabList className={styles.reactTabsTabList}>
        {upcomingReleases.map(release => (
          <Tab
            key={release.title}
            className={styles.reactTabsTab}
            selectedClassName={styles.reactTabsTabSelected}
          >
            Node.js {release.title}
          </Tab>
        ))}
      </TabList>
      {upcomingReleases.map(release => (
        <TabPanel
          key={release.title}
          className={styles.reactTabsTabPanel}
          selectedClassName={styles.reactTabsTabPanelSelected}
        >
          <UpcomingReleasesPanel releases={release.releases} />
        </TabPanel>
      ))}
    </Tabs>
  </div>
);

export default UpcomingReleases;
