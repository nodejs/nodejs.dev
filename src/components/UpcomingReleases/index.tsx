import React from 'react';
import { Tab, TabList, TabPanel, Tabs } from 'react-tabs';
import UpcomingReleasesPanel from './UpcomingReleasesPanel';
import {
  NODE_12_RELEASES,
  NODE_13_RELEASES,
  NODE_14_RELEASES,
  NODE_15_RELEASES,
} from './upcomingReleases';
import './UpcomingReleases.scss';

export default function UpcomingReleases(): JSX.Element {
  return (
    <div className="upcoming-releases">
      <Tabs>
        <TabList>
          <Tab>Node.js 12</Tab>
          <Tab>Node.js 13</Tab>
          <Tab>Node.js 14</Tab>
          <Tab>Node.js 15</Tab>
        </TabList>
        <TabPanel>
          <UpcomingReleasesPanel releases={NODE_12_RELEASES} />
        </TabPanel>
        <TabPanel>
          <UpcomingReleasesPanel releases={NODE_13_RELEASES} />
        </TabPanel>
        <TabPanel>
          <UpcomingReleasesPanel releases={NODE_14_RELEASES} />
        </TabPanel>
        <TabPanel>
          <UpcomingReleasesPanel releases={NODE_15_RELEASES} />
        </TabPanel>
      </Tabs>
    </div>
  );
}
