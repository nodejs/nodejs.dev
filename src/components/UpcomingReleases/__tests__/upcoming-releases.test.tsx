import React from 'react';
import { render } from '@testing-library/react';
import UpcomingReleases from '..';

describe('UpcomingReleases component', (): void => {
  it('renders correctly', (): void => {
    const { container } = render(<UpcomingReleases />);
    expect(container).toMatchSnapshot();
  });
});
