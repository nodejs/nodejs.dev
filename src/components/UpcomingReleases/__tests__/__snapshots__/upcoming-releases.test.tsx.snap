// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpcomingReleases component renders correctly 1`] = `
<div>
  <div
    class="upcoming-releases"
  >
    <div
      class="react-tabs"
      data-tabs="true"
    >
      <ul
        class="react-tabs__tab-list"
        role="tablist"
      >
        <li
          aria-controls="react-tabs-1"
          aria-disabled="false"
          aria-selected="true"
          class="react-tabs__tab react-tabs__tab--selected"
          id="react-tabs-0"
          role="tab"
          tabindex="0"
        >
          Node.js 12
        </li>
        <li
          aria-controls="react-tabs-3"
          aria-disabled="false"
          aria-selected="false"
          class="react-tabs__tab"
          id="react-tabs-2"
          role="tab"
        >
          Node.js 13
        </li>
        <li
          aria-controls="react-tabs-5"
          aria-disabled="false"
          aria-selected="false"
          class="react-tabs__tab"
          id="react-tabs-4"
          role="tab"
        >
          Node.js 14
        </li>
        <li
          aria-controls="react-tabs-7"
          aria-disabled="false"
          aria-selected="false"
          class="react-tabs__tab"
          id="react-tabs-6"
          role="tab"
        >
          Node.js 15
        </li>
        <li
          aria-controls="react-tabs-9"
          aria-disabled="false"
          aria-selected="false"
          class="react-tabs__tab"
          id="react-tabs-8"
          role="tab"
        >
          Node.js 16
        </li>
      </ul>
      <div
        aria-labelledby="react-tabs-0"
        class="react-tabs__tab-panel react-tabs__tab-panel--selected"
        id="react-tabs-1"
        role="tabpanel"
      >
        <div
          class="upcoming-releases__panel"
        >
          <div
            class="upcoming-releases__item--current"
          >
            <img
              alt="hexagon icon"
              src="test-file-stub"
            />
            <p
              class="release-title"
            >
              Current
            </p>
            <p
              class="release-date"
            >
              Released
               
              2019-04-23
            </p>
          </div>
          <div
            class="upcoming-releases__item--lts"
          >
            <img
              alt="hexagon icon"
              src="test-file-stub"
            />
            <p
              class="release-title"
            >
              LTS
            </p>
            <p
              class="release-date"
            >
              Released
               
              2019-10-21
            </p>
          </div>
          <div
            class="upcoming-releases__item--maintenance"
          >
            <img
              alt="hexagon icon"
              src="test-file-stub"
            />
            <p
              class="release-title"
            >
              Maintenance Release
            </p>
            <p
              class="release-date"
            >
              Released
               
              2020-10-20
            </p>
          </div>
          <div
            class="upcoming-releases__item--endoflife upcoming-releases__item--to-be-released"
          >
            <img
              alt="hexagon icon"
              src="test-file-stub"
            />
            <p
              class="release-title"
            >
              End-of-life Release
            </p>
            <p
              class="release-date"
            >
              To be released
               
              2022-04-30
            </p>
          </div>
        </div>
      </div>
      <div
        aria-labelledby="react-tabs-2"
        class="react-tabs__tab-panel"
        id="react-tabs-3"
        role="tabpanel"
      />
      <div
        aria-labelledby="react-tabs-4"
        class="react-tabs__tab-panel"
        id="react-tabs-5"
        role="tabpanel"
      />
      <div
        aria-labelledby="react-tabs-6"
        class="react-tabs__tab-panel"
        id="react-tabs-7"
        role="tabpanel"
      />
      <div
        aria-labelledby="react-tabs-8"
        class="react-tabs__tab-panel"
        id="react-tabs-9"
        role="tabpanel"
      />
    </div>
  </div>
</div>
`;
